FE-8: Discount date validation

tedybear26 opened the pull request:
Also modal's accept button is disabled if there are any errors in the form or the dates.
Created separated schema for discount and payment methods and cleaned up code for that.
Changed a little how the Modal works (per instructions of Marco) and adjusted the components to the changes.

Commit:
Discount percentage between 0 and 100

NicolasBlumetto commented on src/components/modal/Modal.tsx:
```diff
@@ -1,52 +1,59 @@
\ No newline at end of file
+	const { acceptText, children, cancelText, show, title, onClose, onAccept, acceptEnabled } = props;
+
+	const [isOpen, setIsOpen] = useState<boolean>(false);
+
+	useEffect(() => {
+		setIsOpen(show);
+	}, [show]);
+
+	const handleModalClose = useCallback(() => {
+		setIsOpen(false);
+		onClose();
+	}, [onClose]);
+
+	const handleModalAccept = useCallback(() => {
+		setIsOpen(false);
+		onAccept();
+	}, [onAccept]);
+
```
A no ser que me este perdiendo de algo, este estado y el useEffect no son necesarios. Basicamente esta emulando el valor de la prop, se puede usar la prop directamente

NicolasBlumetto replied:
Sugiero

```
import React, { useCallback } from "react";
import { Button, Dialog, Container, Box } from "@mui/material";
import CloseIcon from "@mui/icons-material/Close";
import "./Modal.scss";

export type ModalProps = {
	show: boolean;
	title: string;
	children: React.ReactElement;
	cancelText?: string;
	acceptText?: string;
	onClose: () => void;
	onAccept: () => void;
	acceptEnabled?: boolean;
};

export default function Modal(props: ModalProps): React.ReactElement {
	const { acceptText, children, cancelText, show, title, onClose, onAccept, acceptEnabled } = props;

	const handleModalClose = useCallback(() => {
		onClose();
	}, [onClose]);

	const handleModalAccept = useCallback(() => {
		onAccept();
	}, [onAccept]);

	if (!show) return <></>;
	return (
		<Dialog open={show} onClose={handleModalClose} id="MUIModal">
			<Box className="modal">
				<CloseIcon onClick={handleModalClose} />

				<Container className="wrapper">
					<h3>{title}</h3>
					{children}
				</Container>

				<Container className="button-container">
					<Button variant="outlined" sx={{ marginRight: 2 }} onClick={handleModalClose} id="modalCancel">
						{cancelText ?? "Cancelar"}
					</Button>

					<Button variant="contained" onClick={handleModalAccept} disabled={!acceptEnabled} id="modalAccept">
						{acceptText ?? "Aceptar"}
					</Button>
				</Container>
			</Box>
		</Dialog>
	);
}

```

NicolasBlumetto replied:
Es mas tampoco estoy seguro de que memorizar el handleModalClose y el handleModalAccept aporte mucho porque solo se abre un modal a la vez. Se podria ver y si es asi se pueden sacar esos dos handlers y pasarle directamente onClose y onAccept a los botones y dialog.

tedybear26 replied:
Aplique la sugerencia, sobre lo otro no tengo mucha idea.

NicolasBlumetto reviewed (changes requested):
Revisar lo que puse sobre el Modal.  Tambien los scss me parece estan tabualdos pero idk. 
Por lo demas looks good

NicolasBlumetto commented on src/components/modal/Modal.tsx:
```diff
@@ -1,52 +1,59 @@
+					<h3>{title}</h3>
+					{children}
+				</Container>
+
+				<Container className="button-container">
```
Agregarle `sx={{ display: "flex" }}>` al button container. Por defecto el container viene con display:block y hay que pisarlo para que se le apliquen los cambios de flexbox

tedybear26 replied:
Listo

marcobaldi97 commented on src/pages/student/components/AdministrativeInfo/DiscountsSection/DiscountsSection.tsx:
```diff
@@ -4,14 +4,14 @@ import * as Models from "../../../../../core/Models";
 import AddCircleOutlineIcon from "@mui/icons-material/AddCircleOutline";
 
-import schema from "../../../schema.json";
+import schema from "./schema.json";
```
Por qué usar este schema en vez del general? La gracia es que compartan validaciones y errores. En todo caso meter los cambios en el general y usar ese.

tedybear26 replied:
Esto es así porque el form del descuento no edita directamente al alumno, tiene un boton de agregar y eso en el dialogo. Si lo hacia con el schema general tenia que hacer unas chanchadas porque el form me devuelve la data en una estructura mucho mas grande (ej me devuelve administrative-info { discount: { la data que quiero}}

marcobaldi97 replied:
Personalmente me sigue pareciendo un tradeoff razonable si tenes que meter data.administrative_info.discount en un par de lineas antes que tener que mantener dos schemas en vez de uno

tedybear26 replied:
Ok, yo deshice eso como estaba porque lo habiamos comentado con Nico, si queres lo vuelvo para atrás

marcobaldi97 commented on src/pages/student/components/AdministrativeInfo/PaymentMethodsSection/PaymentMethodSection.tsx:
```diff
@@ -4,14 +4,14 @@ import * as Models from "../../../../../core/Models";
 import AddCircleOutlineIcon from "@mui/icons-material/AddCircleOutline";
 
-import schema from "../../../schema.json";
+import schema from "./schema.json";
```
Lo mismo que en el otro caso: Usar el schema general, modificarlo de ser necesario.

tedybear26 replied:
La misma respuesta.

marcobaldi97 reviewed (changes requested):
@tedybear26 :  Hacer los cambios del schema.

marcobaldi97 commented:
<img width="1040" alt="Screen Shot 2022-09-30 at 17 35 12" src="https://user-images.githubusercontent.com/60588169/193352314-7eb14252-f9e2-4b7b-a643-5cb4ba5cdd53.png">
No se si le ando ladrando al perro correcto pero no me deja usar la cruz de arriba, ver imagen

Nadia-Martinez commented:
> <img alt="Screen Shot 2022-09-30 at 17 35 12" width="1040" src="https://user-images.githubusercontent.com/60588169/193352314-7eb14252-f9e2-4b7b-a643-5cb4ba5cdd53.png">
> 
> No se si le ando ladrando al perro correcto pero no me deja usar la cruz de arriba, ver imagen

A mi me anduvo, lo que si vería si podemos ponerle un `cursor: pointer` a los botones de "+" para que se vean con manito

Commit:
Use student schema for payment and discount

# Conflicts:
#	src/pages/student/schema.json

Commit:
fixed forms in discount and payment pop ups

Commit:
Discount date validation

Also modal's accept button is disabled if there are any errors in the form or the dates.

# Conflicts:
#	src/pages/student/components/AdministrativeInfo/DiscountsSection/DiscountsSection.tsx
#	src/pages/student/components/AdministrativeInfo/DiscountsSection/ui.json

# Conflicts:
#	src/components/modal/Modal.tsx
#	src/pages/student/components/AdministrativeInfo/DiscountsSection/DiscountsSection.tsx

Commit:
Changes to modal. Schema for Payment Method

# Conflicts:
#	src/pages/student/components/AdministrativeInfo/DiscountsSection/DiscountsSection.tsx
#	src/pages/student/components/AdministrativeInfo/PaymentMethodsSection/ui.json

# Conflicts:
#	src/components/modal/Modal.tsx
#	src/pages/student/components/AdministrativeInfo/DiscountsSection/DiscountsSection.tsx
#	src/pages/student/components/AdministrativeInfo/PaymentMethodsSection/PaymentMethodSection.tsx

Commit:
Requested changes

# Conflicts:
#	src/components/modal/Modal.tsx

Commit:
Typography change

# Conflicts:
#	src/components/modal/Modal.tsx

tedybear26 commented:
@NicolasBlumetto REvisa si te parece bien el cambio de volver al schema general y si te parece bien dale approve.
